x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/TimeseriesLifecycleType.java

```diff
@@ -363,6 +365,46 @@ static void validateAllSearchableSnapshotActionsUseSameRepository(Collection<Pha
         }
     }
 
+    /**
+     * Validates that phases don't configure a min_age that is smaller than a previous phase (which can be confusing to users)
+     */
+    public static String validateMonotonicallyIncreasingPhaseTimings(Collection<Phase> phases) {
+        List<String> errors = new ArrayList<>();
+
+        // Loop through all phases in order, for each phase with a min_age
+        // configured, look at all the future phases to see if their ages are
+        // >= the configured age. A min_age of 0 means that the age was not
+        // configured, so we don't validate it.
+        for (int i = 0; i < ORDERED_VALID_PHASES.size(); i++) {
+            String phaseName = ORDERED_VALID_PHASES.get(i);
+            // Check if this phase is present with a configured min_age
+            Optional<Phase> maybePhase = phases.stream()
+                .filter(p -> phaseName.equals(p.getName()))
+                .filter(p -> p.getMinimumAge() != null && p.getMinimumAge().equals(TimeValue.ZERO) == false)
+                .findFirst();
+
+            if (maybePhase.isPresent()) {
+                Phase phase = maybePhase.get();
+                TimeValue phaseMinAge = phase.getMinimumAge();
+                Set<String> followingPhases = new HashSet<>(ORDERED_VALID_PHASES.subList(i + 1, ORDERED_VALID_PHASES.size()));
+                Set<Phase> phasesWithBadAges = phases.stream()
+                    .filter(p -> followingPhases.contains(p.getName()))
```

Comment:
I believe that would always be true, but maybe I'm misunderstanding the intent here?

Reply:
So here we are looping through the phases, if we're on the "warm" phase then the `followingPhases` will be `cold, frozen, delete`. Then, this takes the phases configured in the policy (let's say it has hot, warm, frozen, and delete) and filters them to return a stream with "frozen" and "delete" in it because there is no "cold" phase.

Hopefully that makes sense?

Reply:
That makes sense, but I thought the sublist above doest that already?
```
ORDERED_VALID_PHASES.subList(i + 1, ORDERED_VALID_PHASES.size())
```

Which I'm just seeing is placed in a `HashSet` so the order might be lost? I think we should keep an `ArrayList` above?
```
List<String> followingPhases = new ArrayList<>(ORDERED_VALID_PHASES.subList(i + 1, ORDERED_VALID_PHASES.size()));
```

Reply:
Ah, I finally got there. I mixed-up the collections.

So I think the order can be indeed lost because we don't expect the `validateMonotonicallyIncreasingPhaseTimings` to be called with the list of ordered phases (and it is not https://github.com/elastic/elasticsearch/pull/70089/files#diff-dbca592978c7f9e3f747487f76b2e7327bdcac2fa635a23949aa35fba7411bfbR65) but we then iterate over the `phases` collection we received as if it is ordered - https://github.com/elastic/elasticsearch/pull/70089/files#diff-5964cde83b3b053cf693079cb18a73f4a5743c59590f09fda7dd268613098ce9R390

This is where the mix was in my mind. I think if we keep the order in the `followingPhases` collection and iterate over it we can overcome this:

eg. 

```
followingPhases.stream().filter(p -> p.getMinimumAge() != null && p.getMinimumAge().equals(TimeValue.ZERO) == false).filter(p -> p.getMinimumAge().compareTo(phaseMinAge) < 0)
```


```

Reply:
> So I think the order can be indeed lost because we don't expect the validateMonotonicallyIncreasingPhaseTimings to be called with the list of ordered phases (and it is not #70089 (files)) but we then iterate over the phases collection we received as if it is ordered

We don't iterate over the phases collection received though, instead we iterate over `ORDERED_VALID_PHASES`, picking out the phases from the provided collection one-by-one

`followingPhases` doesn't need to be ordered because it is a subset of the phases occurring *after* the `maybePhase`, so if `phase` were `warm` then `followingPhases` would be `cold, frozen, delete`, and the order for those doesn't matter because they are all compared independently to the timings for the "warm" phase.

Hopefully that was clear? I'm probably not explaining it in the clearest way.

Reply:
Thanks for the clarification. I was thrown by the `phases.stream()` line which is iterating over the possibly unordered parameter list, however the `.filter(p -> followingPhases.contains(p.getName()))` filtering makes sure we do only process the following phases, and you're right the order at that stage does not matter anymore. 

Thanks for bearing with me!